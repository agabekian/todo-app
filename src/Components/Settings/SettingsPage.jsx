import React, { useContext } from 'react';
import { SettingsContext } from '/src/context/Settings.jsx';
import { Container } from '@mantine/core';

const SettingsPage = () => {
    const { hideCompleted, itemsToShow, setHideCompleted, setItemsToShow } = useContext(SettingsContext);

    const handleHideCompletedChange = () => {
        setHideCompleted((prev) => !prev);
    };

    const handleItemsToShowChange = (event) => {
        setItemsToShow(parseInt(event.target.value));
    };

    return (
        <Container size="md" style={{ padding: '2rem' }}>
            <div>
                <label>
                    Hide Completed:
                    <input type="checkbox" checked={hideCompleted} onChange={handleHideCompletedChange} />
                </label>
            </div>
            <div>
                <label>
                    Items to Show:
                    <select value={itemsToShow} onChange={handleItemsToShowChange}>
                        {[...Array(20)].map((_, index) => (
                            <option key={index + 1} value={index + 1}>
                                {index + 1}
                            </option>
                        ))}
                    </select>
                </label>
            </div>
        </Container>
    );
};

export default SettingsPage;
