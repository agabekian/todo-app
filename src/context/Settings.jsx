import React, {useState} from 'react';

export const SettingsContext = React.createContext();

function SettingsProvider(props) {
    const [itemsToShow] = useState(5);
    const [hideCompleted] = useState(true);
    const [defaultSort] = useState('difficulty'); // for optional stretch

    // const toggleMode = () => setMode(mode === 'light' ? 'dark' : 'light');
    return (
        <SettingsContext.Provider value={{itemsToShow, hideCompleted}}>
            {props.children}
        </SettingsContext.Provider>
    );
}

export default SettingsProvider;
