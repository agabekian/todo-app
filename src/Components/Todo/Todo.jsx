import React, { useEffect, useState, useContext } from 'react';
import { SettingsContext } from '/src/context/Settings.jsx';
import useForm from '../../hooks/form';
import { v4 as uuid } from 'uuid';
import { Container, Group, Pagination } from '@mantine/core';
import '@mantine/core/styles.css';
import Form from "../Form/Form.jsx";
import TaskCard from "../TaskCard.jsx";
import Auth from "../auth/auth.jsx";

const Todo = () => {
    const settings = useContext(SettingsContext);
    const [defaultValues] = useState({
        difficulty: 4,
    });
    const [displayList, setDisplayList] = useState([]);
    const [incomplete, setIncomplete] = useState([]);
    const { handleChange, handleSubmit } = useForm(addItem, defaultValues);
    const itemsToShow = settings.itemsToShow;
    const [list, setList] = useState(() => {
        const savedList = localStorage.getItem('todoList');
        return savedList ? JSON.parse(savedList) : [];
    });
    const [currentPage, setCurrentPage] = useState(() => {
        const savedPage = localStorage.getItem('currentPage');
        return savedPage ? JSON.parse(savedPage) : 1;
    });
    const [hideCompleted, setHideCompleted] = useState(settings.hideCompleted);

    function addItem(item) {
        item.id = uuid();
        item.complete = false;
        setList([...list, item]);
    }

    function toggleComplete(id) {
        const updatedList = list.map(item => {
            if (item.id === id) {
                item.complete = !item.complete;
            }
            return item;
        });

        setList(updatedList);
    }

    function deleteItem(id) {
        const items = list.filter(item => item.id !== id);
        setList(items);
    }

    useEffect(() => {
        const filteredList = hideCompleted ? list.filter(item => !item.complete) : list;
        const totalPages = Math.ceil(filteredList.length / itemsToShow);
        setDisplayList(filteredList.slice((currentPage - 1) * itemsToShow, currentPage * itemsToShow));

        setIncomplete(filteredList.filter(item => !item.complete).length);

        localStorage.setItem('todoList', JSON.stringify(list));
        localStorage.setItem('currentPage', JSON.stringify(currentPage));
        document.title = `To Do List: ${incomplete}`;
    }, [list, currentPage, itemsToShow, hideCompleted]);

    useEffect(() => {
        setHideCompleted(settings.hideCompleted);
    }, [settings.hideCompleted]);

    useEffect(() => {
        // Filter the list based on hideCompleted setting
        const filteredList = hideCompleted ? list.filter(item => !item.complete) : list;

        // Calculate incomplete count
        const incompleteCount = filteredList.filter(item => !item.complete).length;
        setIncomplete(incompleteCount);

        // Calculate total pages
        const totalPages = Math.ceil(filteredList.length / itemsToShow);

        // Update displayList with proper pagination
        const startIndex = (currentPage - 1) * itemsToShow;
        const endIndex = Math.min(startIndex + itemsToShow, filteredList.length);
        const displayItems = filteredList.slice(startIndex, endIndex);
        setDisplayList(displayItems);

        // Update localStorage and document title
        localStorage.setItem('todoList', JSON.stringify(list));
        localStorage.setItem('currentPage', JSON.stringify(currentPage));
        document.title = `To Do List: ${incomplete}`;
    }, [list, currentPage, itemsToShow, hideCompleted]);



    function handlePageChange(page) {
        setCurrentPage(page);
    }

    return (
        <Container size="xl" style={{ padding: '2rem' }}>
            <p data-testid="todo-component">{incomplete}</p>
            <Group position="apart" align="flex-start">
                <Auth capability={'create'}>
                    <Form handleChange={handleChange}
                          handleSubmit={handleSubmit}
                          difficulty={defaultValues.difficulty} />
                </Auth>

                <TaskCard list={displayList}
                          toggleComplete={toggleComplete}
                          deleteItem={deleteItem} />
            </Group>

            <Pagination
                style={{ position: 'fixed', bottom: '20px', right: '20px', zIndex: '100' }}
                total={Math.ceil(incomplete / itemsToShow)}
                page={currentPage}
                onChange={handlePageChange}
                data-testid="pagination-component"
            />
        </Container>
    );
};

export default Todo;
