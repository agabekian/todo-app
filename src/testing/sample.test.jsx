import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { describe, it, expect } from 'vitest';
import React from 'react';
import { SettingsContext } from '/src/context/Settings.jsx'; // Import the SettingsContext
import Todo from '../Components/Todo/Todo.jsx'; // Adjust the import path as necessary

describe('Todo Component', () => {
  // Mock the SettingsContext value
  const settingsValue = {
    itemsToShow: 4, // Define itemsToShow value for testing
    hideCompleted: false // You can add other properties if needed
  };

  it('renders the component and matches the snapshot', () => {
    const { asFragment } = render(
        <SettingsContext.Provider value={settingsValue}> {/* Provide the mock value */}
          <Todo />
        </SettingsContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('adds a new to-do item', () => {
    render(
        <SettingsContext.Provider value={settingsValue}>
          <Todo />
        </SettingsContext.Provider>
    );

    fireEvent.change(screen.getByPlaceholderText(/Item Details/i), { target: { value: 'New Task' } });
    fireEvent.change(screen.getByPlaceholderText(/Assignee Name/i), { target: { value: 'John Doe' } });
    fireEvent.click(screen.getByText(/Add Item/i));

    expect(screen.getByText('New Task')).toBeInTheDocument();
    expect(screen.getByText(/Assigned to: John Doe/i)).toBeInTheDocument();
    expect(screen.getAllByText(/Difficulty:/i)).toHaveLength(1);
  });

  it('toggles completion status of a to-do item', () => {
    render(
        <SettingsContext.Provider value={settingsValue}>
          <Todo />
        </SettingsContext.Provider>
    );

    fireEvent.change(screen.getByPlaceholderText(/Item Details/i), { target: { value: 'New Task' } });
    fireEvent.change(screen.getByPlaceholderText(/Assignee Name/i), { target: { value: 'John Doe' } });
    fireEvent.click(screen.getByText(/Add Item/i));

    const completeDiv = screen.getByText(/Complete: false/i);
    expect(completeDiv).toBeInTheDocument();

    fireEvent.click(completeDiv);
    expect(screen.getByText(/Complete: true/i)).toBeInTheDocument();
  });
//############# DELETE ############# not implementd so not tested YET
//   it('deletes a to-do item', () => {
//     render(
//         <SettingsContext.Provider value={settingsValue}>
//           <Todo />
//         </SettingsContext.Provider>
//     );
//
//     // Add a new to-do item
//     fireEvent.change(screen.getByPlaceholderText(/Item Details/i), { target: { value: 'New Task' } });
//     fireEvent.change(screen.getByPlaceholderText(/Assignee Name/i), { target: { value: 'John Doe' } });
//     fireEvent.click(screen.getByText(/Add Item/i));
//
//     // Verify that the new to-do item is present
//     expect(screen.getByText('New Task')).toBeInTheDocument();
//
//     // Mark the to-do item as complete
//     fireEvent.click(screen.getByText(/Complete: false/i));
//     fireEvent.click(screen.getByText(/Complete: true/i));
//
//     // Debugging: Log the document body to see if the item is still present
//     console.log(document.body.innerHTML);
//
//     // Verify that the to-do item is no longer present (deleted)
//     expect(screen.queryByText('New Task')).not.toBeInTheDocument();
//   });


  it('limits the display of to-do items to 4', () => {
    render(
        <SettingsContext.Provider value={settingsValue}>
          <Todo />
        </SettingsContext.Provider>
    );

    for (let i = 1; i <= 5; i++) {
      fireEvent.change(screen.getByPlaceholderText(/Item Details/i), { target: { value: `Task ${i}` } });
      fireEvent.change(screen.getByPlaceholderText(/Assignee Name/i), { target: { value: `Person ${i}` } });
      fireEvent.click(screen.getByText(/Add Item/i));
    }

    const tasks = screen.getAllByText(/Task/i);
    expect(tasks).toHaveLength(4);
  });
});
