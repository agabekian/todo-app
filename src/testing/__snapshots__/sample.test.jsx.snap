// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Todo Component > renders the component and matches the snapshot 1`] = `
<DocumentFragment>
   
  <header
    data-testid="todo-header"
  >
    <h1
      data-testid="todo-h1"
    >
      To Do List: 0 items pending
    </h1>
  </header>
  <form>
    <h2>
      Add To Do Item
    </h2>
    <label>
      <span>
        To Do Item
      </span>
      <input
        name="text"
        placeholder="Item Details"
        type="text"
      />
    </label>
    <label>
      <span>
        Assigned To
      </span>
      <input
        name="assignee"
        placeholder="Assignee Name"
        type="text"
      />
    </label>
    <label>
      <span>
        Difficulty
      </span>
      <input
        max="5"
        min="1"
        name="difficulty"
        type="range"
        value="4"
      />
    </label>
    <label>
      <button
        type="submit"
      >
        Add Item
      </button>
    </label>
  </form>
</DocumentFragment>
`;
